machine/v1beta1/types_machine.go

```diff
@@ -194,6 +199,38 @@ type MachineSpec struct {
 	ProviderID *string `json:"providerID,omitempty"`
 }
 
+// LifecycleHooks allow users to pause operations on the machine at
+// certain prefedined points within the machine lifecycle.
+type LifecycleHooks struct {
+	// PreDrain hooks prevent the machine from being drained.
+	// This also blocks further lifecycle events, such as termination.
+	// +optional
+	PreDrain []LifecycleHook `json:"preDrain,omitempty"`
+
+	// PreTerminate hooks prevent the machine from being terminated.
+	// PreTerminate hooks be actioned after the Machine has been drained.
+	// +optional
+	PreTerminate []LifecycleHook `json:"preTerminate,omitempty"`
+}
+
+// LifecycleHook represents a single instance of a lifecycle hook
+type LifecycleHook struct {
+	// Name defines a unique name for the lifcycle hook.
+	// The name should be unique and descriptive, ideally 1-3 words, in camelCase with no spaces.
+	// Names must be unique and should only be managed by a single entity.
+	// +kubebuilder:validation:Pattern:="[A-Za-z]+"
+	// +required
+	Name string `json:"name"`
+
+	// Owner defines the owner of the lifecycle hook.
+	// This should be descriptive enough so that users can identify
+	// who/what is responsible for blocking the lifecycle.
+	// This could be the name of a controller (e.g. clusteroperator/etcd)
+	// or an administrator managing the hook.
+	// +required
+	Owner string `json:"owner"`
```

Comment:
can this value be empty? if not can we add validation for it?

Reply:
Value cannot be empty, it is expected to always be populated in each entry, though with what we don't specify. I think this can be free form text. Not sure what validation you had in mind?

Reply:
maybe some minimal length if it makes sense here

Reply:
I think a minimum length is hard to quantify, we could put 5 characters, but then what if the owner is `etcd`, or something equivalently short. I think we should trust operators to put something useful within this field

Reply:
how about min==3 max==1024?

Reply:
Do we typically have a maximum length for Condition Types? Since the name becomes part of the condition type, perhaps it's sensible to even limit at 256

Reply:
> Do we typically have a maximum length for Condition Types? Since the name becomes part of the condition type, perhaps it's sensible to even limit at 256

We do, see comment below.